pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -84,14 +86,54 @@ func NewOperatorStepResolver(lister operatorlister.OperatorLister, client versio
 	return stepResolver
 }
 
+// replacementChainFailed returns true if the latest entry in a chain of operators is in the failed state.
+func replacementChainFailed(csv *v1alpha1.ClusterServiceVersion, csvToReplacement map[string]*v1alpha1.ClusterServiceVersion) bool {
+	if csv == nil {
+		return false
+	}
+	if csv.Status.Phase == v1alpha1.CSVPhaseFailed {
+		return true
+	}
+	return replacementChainFailed(csvToReplacement[csv.GetName()], csvToReplacement)
```

Comment:
The last CSV in a replacement chain shouldn't have a replacement, so doesn't the base-case `csv == nil` always occur, even if the csv isn't failed (order of base-case issue)?